esmvalcore/experimental/config/_config_validators.py

```diff
@@ -0,0 +1,284 @@
+# the the 'Python Software Foundation License'
+# (https://www.python.org/psf/license)
+def _make_type_validator(cls, *, allow_none=False):
+    """Construct a type validator for `cls`.
+
+    Return a validator that converts inputs to *cls* or raises (and
+    possibly allows ``None`` as well).
+    """
+    def validator(inp):
+        looks_like_none = isinstance(inp, str) and (inp.lower() == "none")
+        if (allow_none and (inp is None or looks_like_none)):
+            return None
+        try:
+            return cls(inp)
+        except ValueError as err:
+            if isinstance(cls, type):
+                raise ValueError(
+                    f'Could not convert {repr(inp)} to {cls.__name__}'
+                ) from err
+            raise
+
+    validator.__name__ = f"validate_{cls.__name__}"
+    if allow_none:
+        validator.__name__ += "_or_None"
+    validator.__qualname__ = (validator.__qualname__.rsplit(".", 1)[0] + "." +
+                              validator.__name__)
+    return validator
+
+
+# The code for this function was taken from matplotlib (v3.3) and modified
+# to fit the needs of ESMValCore. Matplotlib is licenced under the terms of
+# the the 'Python Software Foundation License'
+# (https://www.python.org/psf/license)
+@lru_cache()
+def _listify_validator(scalar_validator,
+                       allow_stringlist=False,
+                       *,
+                       n_items=None,
+                       docstring=None):
+    """Apply the validator to a list."""
+    def func(inp):
+        if isinstance(inp, str):
+            try:
+                inp = [
+                    scalar_validator(val.strip()) for val in inp.split(',')
+                    if val.strip()
+                ]
+            except Exception:
+                if allow_stringlist:
+                    # Sometimes, a list of colors might be a single string
+                    # of single-letter colornames. So give that a shot.
+                    inp = [
+                        scalar_validator(val.strip()) for val in inp
+                        if val.strip()
+                    ]
+                else:
+                    raise
+        # Allow any ordered sequence type -- generators, np.ndarray, pd.Series
+        # -- but not sets, whose iteration order is non-deterministic.
+        elif isinstance(inp,
+                        Iterable) and not isinstance(inp, (set, frozenset)):
+            # The condition on this list comprehension will preserve the
+            # behavior of filtering out any empty strings (behavior was
+            # from the original validate_stringlist()), while allowing
+            # any non-string/text scalar values such as numbers and arrays.
+            inp = [
+                scalar_validator(val) for val in inp
+                if not isinstance(val, str) or val
+            ]
+        else:
+            raise ValueError(
```

Comment:
Should we raise custom error instead of generic ValueError here (and the rest of this code)?